### DISCUSSION TITLE
Debugging segfaults "No module named qutebrowser.__main__;..."

### DISCUSSION THREAD

**Discussion by steveb-123** [qutebrowser/qutebrowser#6226] [Q&A] [answered]
**Background:**
I'm having no end of crashes of qutebrowser that are seemingly driven by random specific webpages. Sometimes a specific page will load, another day it will cause immediate crashes (even persistent when I reload qutebrowser, and the tab is still open -  i have  to rapidly delete the tab or qutebrowser re-crashes). [version](https://paste.the-compiler.org/view/c164cfda). Websites which have causes this have included outlook web access, springer journal sites, google maps and more. I'm using an infuriating work laptop which is locked at ubuntu 18.04 which I would not be surprised is the root of all my problems.

To try and solve my problems I've installed the debugging packages recommended in the "getting a stacktrace.." page. However, when I try to run:

```
 gdb -ex r --args $(readlink -f $(which python3)) -m qutebrowser --debug --temp-basedir
```

I get

```
...
Starting program: /usr/bin/python3.6 -m qutebrowser --debug --temp-basedir
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
/usr/bin/python3.6: No module named qutebrowser.__main__; 'qutebrowser' is a package and cannot be directly executed
[Inferior 1 (process 10854) exited with code 01]
```

Any suggestion regarding my debugging issue or even my original issue?

thanks :)

**Comment by steveb-123** [comment]
To clarify, these are the debugging packages I have:

```
python3-pyqt5-dbg is already the newest version (5.10.1+dfsg-1ubuntu2).
python3-pyqt5.qtwebengine-dbg is already the newest version (5.10.1+dfsg-1ubuntu2).
python3-dbg is already the newest version (3.6.7-1~18.04).
libqt5webenginecore5-dbgsym is already the newest version (5.9.5+dfsg-0ubuntu2).
libqt5webengine5-dbgsym is already the newest version (5.9.5+dfsg-0ubuntu2).
```

I hope those version numbers are decipherable

**Comment by The-Compiler** [comment, marked as answer]
Cany ou try using `.venv/bin/python3` instead of `$(readlink -f $(which python3))` and see if that helps?

Also note that you installed qutebrowser in a virtualenv, so Ubuntu's debug packages won't help you there - unfortunately, there's no easy way to get debug symbols for a virtualenv-installed PyQt at the moment.

Given that your QtWebEngine version is slightly outdated and QtWebEngine 5.15.0 had various bugs, the first thing I'd try is [running `mkvenv.py --update`](https://github.com/qutebrowser/qutebrowser/blob/master/doc/install.asciidoc#updating).

**Reply by steveb-123** [reply]
Thanks for the help!

I tried previously both QtWebEngine 5.15.3 and 5.15.0. I thought at first that this downgrade has fixed my issues, but they always come back again.


Switching to ``.../qutebrowser/`` and running

```
gdb -ex r --args .venv/bin/python3 -m qutebrowser --debug --temp-basedir
```

got me some debugging text.

However, it seems that running with `--temp-basedir` solves my problems in the short term. What does this actually do? Is this an indication that my config is broken? Is there some directory which I might be able to delete to solve my problems.

When I run the above without --temp-basedir I get the crashes again, reproducibly.

```
11:39:28 DEBUG    webview    tabbedbrowser:_on_title_changed:758 Changing title for idx 0 to 'https://www.nature.com/articles/s41598-019-40331-0'
11:39:28 DEBUG    misc       eventfilter:eventFilter:63 <qutebrowser.browser.webengine.webview.WebEngineView object at 0x7fff8c0ca828>: removed child <PyQt5.QtCore.QObject object at 0x7fff8c077678>
11:39:28 DEBUG    webview    tabbedbrowser:_on_title_changed:758 Changing title for idx 0 to 'E-syt1 Re-arranges STIM1 Clusters to Stabilize Ring-shaped ER-PM Contact Sites and Accelerate Ca 2+ Store Replenishment | Scientific Reports'
11:39:28 DEBUG    js         shared:javascript_log_message:155 [https://www.nature.com/articles/s41598-019-40331-0:0] An <img> element was lazyloaded with loading=lazy, but had no dimensions specified. Specifying dimensions improves performance. See https://crbug.com/954323
11:39:28 DEBUG    js         shared:javascript_log_message:155 [https://www.nature.com/static/js/shared-es6-bundle.8922d7aab1.js:1] executing entitlements
11:39:28 DEBUG    js         shared:javascript_log_message:155 [https://www.nature.com/static/js/shared-es6-bundle.8922d7aab1.js:1] initing EntitlementsManager
11:39:28 DEBUG    js         shared:javascript_log_message:155 [https://www.nature.com/static/js/shared-es6-bundle.8922d7aab1.js:1] EntitlementIdPNatureLoader called
11:39:28 DEBUG    js         shared:javascript_log_message:155 [https://www.nature.com/static/js/shared-es6-bundle.8922d7aab1.js:1] loadIdPNatureVerify called
11:39:28 DEBUG    webview    browsertab:_set_load_status:994 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=0 url='https://www.nature.com/articles/s41598-019-40331-0'>: LoadStatus.success_https
11:39:28 DEBUG    signals    signalfilter:_filter_signals:88 emitting: cur_load_status_changed(<LoadStatus.success_https: 3>) (tab 0)
11:39:29 DEBUG    webview    browsertab:_on_navigation_request:1055 navigation request: url https://gum.criteo.com/syncframe?topUrl=www.nature.com#%7B%22bundle%22:%7B%22origin%22:0,%22value%2…, type Type.other, is_main_frame False
11:39:29 DEBUG    js         shared:javascript_log_message:155 [https://www.nature.com/articles/s41598-019-40331-0:0] A cookie associated with a cross-site resource at http://www.facebook.com/ was set without the `SameSite` attribute. A future release of Chrome will only deliver cookies with cross-site requests if they are set with `SameSite=None` and `Secure`. You can review cookies in developer tools under Application>Storage>Cookies and see more details at https://www.chromestatus.com/feature/5088147346030592 and https://www.chromestatus.com/feature/5633521622188032.
11:39:29 DEBUG    js         shared:javascript_log_message:155 [https://www.nature.com/articles/s41598-019-40331-0:0] A cookie associated with a cross-site resource at https://www.google.com/ was set without the `SameSite` attribute. A future release of Chrome will only deliver cookies with cross-site requests if they are set with `SameSite=None` and `Secure`. You can review cookies in developer tools under Application>Storage>Cookies and see more details at https://www.chromestatus.com/feature/5088147346030592 and https://www.chromestatus.com/feature/5633521622188032.
[New Thread 0x7fff8dffb700 (LWP 18399)]
[New Thread 0x7fffbd7fa700 (LWP 18401)]
[New Thread 0x7fff8cff9700 (LWP 18400)]
[New Thread 0x7fffbffff700 (LWP 18402)]
Thread 19 "Chrome_IOThread" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff8effd700 (LWP 18324)]
0x00007fffe79b8ab7 in ?? () from /home/*******/qutebrowser/.venv/lib/python3.6/site-packages/PyQt5/Qt/lib/libQt5WebEngineCore.so.5
```

I don't know if there is anything useful there... is that enough of a tail?

Cheers

**Reply by The-Compiler** [reply]
The crash being in `Chrome_IOThread` hints at #5634. Do they go away if you move `~/.local/share/qutebrowser/webengine/Service Worker` away (while qutebrowser is closed)?

**Reply by steveb-123** [reply]
`mv`ing that directory sorted my issues at least for now.

I will set your `qt.workarounds.remove_service_workers` to `true` having read the issue you cited, by adding

```
c.qt.workarounds.remove_service_workers = True
```

to my config.

You might want to add something to the documentation/troubleshooting/stacktrace page referencing this as I couldn't find that issue so easily. Diagnosis is pretty tough at the end user level and it is originally dated quite far back in git issues. I happened on the issue now as I recently switched to venv installation, it mustn't have been a problem on the archaic ubuntu 18.04 qutebrowser version.

Thanks for your help with this!

**Reply by The-Compiler** [reply]
Well, ideally I'd still like to find out what's causing it rather than everyone just enabling that workaround (which might cause certain website data not to be persisted and also increases qutebrowser's startup time).

The as-of-yet unsolved question is why the data in that directory gets corrupted. I've never been able to reproduce the issue myself.

**Reply by steveb-123** [reply]
I understand completely, I thought that might've been your reasoning.

I don't know how prevalent this issue is, but for me it was crippling my ability to use qutebrowser, so I worry that you might lose some users if its too hard to find the fix even in the interim. 

Regarding peculiarities of my setup which I am aware of, incase it helps one day to reproduce/understand: 

*   I am on public institutional wifi network which masks certain trackers etc: as an example, the links in a google search for "shopping" sites which redirect through those anonymous trackers are blocked. This network causes issues randomly with some sites.
*   I have an infuriating old ubuntu distribution version (18.04) which I am blocked from changing, so many packages are ancient (a painful experience coming from arch on my own systems).

If I get a chance at some point I might try and use only 4G internet after removing the service worker folder, to see if it driven by the network.
